### PR TITLE
[WIP] Still working through CH3

### PR DESCRIPTION
**Pull request by JosephTLucas**

### CODE REVIEW

File: src/Chapter3.hs [hunk 1]
```diff
@@ -373,7 +377,22 @@ after the fight. The battle has the following possible outcomes:
    doesn't earn any money and keeps what they had before.
 
 -}
-
+data Knight = MkKnight
+    { khealth    :: Int
+    , kattack    :: Int
+    , kgold      :: Int
+    }
+--type Monster = Knight --is there a way to sublcass like this while still getting a MkMonster? How about preventing negative values in definitions for health/attack/gold?
```
Comment (JosephTLucas) [inline src/Chapter3.hs:385]: I noticed monster and knight have the same attributes.  This seemed like a chance to alias, but then the constructor seems like the wrong type.  Is there a way to remedy that?  Do I create a "parent class" that both knight and monster inherit from?

File: src/Chapter3.hs [hunk 2]
```diff
@@ -373,7 +377,22 @@ after the fight. The battle has the following possible outcomes:
...
+data Monster = MkMonster
+    { mhealth   :: Int
+    , mattack   :: Int
+    , mgold     :: Int
+    }
+fight :: Monster -> Knight -> Int
+fight m k
+    | kattack k >= mhealth m = kgold k + mgold m
```
Comment (JosephTLucas) [inline src/Chapter3.hs:393]: Is there an equivalent to "dot notation" (`k.attack`)?  Having both `kattack` and `mattack` felt clunky.

File: src/Chapter3.hs [hunk 3]
```diff
@@ -479,6 +505,10 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living people inside
 -}
+data City
```
Comment (JosephTLucas) [inline src/Chapter3.hs:508]: Does this meet the spec?  With this setup, I was having trouble writing the required functions.  I wasn't sure how to address the mutual exclusivity of the `Church` and `Library` and the fact that houses can only have a set number of people (enum?).

**Comment by JosephTLucas** [comment]
summoning @vrom911 and @chshersh for some sage wisdom.

File: src/Chapter3.hs (hunk 2 above)
Comment (vrom911) [reply src/Chapter3.hs:393]: Yes, I understand, how natural such syntax feels. And the Haskel community feels the necessity in this too.
There is a compiler proposal – `RecordDotSyntax` – that would make such syntax possible, but at the moment it doesn't work out of the box.

There is another way to work with the newtypes, that is preferable (at least in Kowainik) at the moment. It is also an additional compiler feature but could be useful in such situations. It is called `RecordWildCards`. You can read more about it in @chshersh blog post: https://kodimensional.dev/recordwildcards

In short, it gives you the ability to use names of the fields as field values (instead of the normal functions):

It could looked like this:

```haskell
{-# LANGUAGE RecordWildCards #-}
fight :: Monster -> Knight -> Int
fight Monster{..} Knight{..}
    | kattack >= mhealth = kgold + mgold
...
```

File: src/Chapter3.hs [hunk 4]
```diff
@@ -459,7 +478,14 @@ and provide more flexibility when working with data types.
...
 -}
-
+data Meal
+    = Breakfast
+    | Second_breakfast
+    | Elevenses
+    | Luncheon
+    | Afternoon_tea
+    | Dinner
+    | Supper
```
Comment (vrom911) [inline src/Chapter3.hs:488]: Wonderful schedule! 😋 ☕

File: src/Chapter3.hs (hunk 3 above)
Comment (vrom911) [reply src/Chapter3.hs:508]: The idea is in the right direction. Although there are a few aspects that need to be done 🙂 

Although this totally makes sense to have a sum type (as you was doing in here), it would be much easier to have `City` as a product type.

Let's see how it fits:

First, we need to decide with the castle. Let's look at the task:

> Wall, but only if the city has a castle
> -- Optional castle with a name (as 'String')

That means that walls are very connected to the castle, and they could not exist separately. So we have three options for the castle:

 1. Castle with a name
 2. Castle with the name + walls
 3. No castle

This looks like a sum type!:

```haskell
data Castle 
    = Castle String
    | WallCastle String
    | None
```

Note that your idea with castle suits here perfectly as well. But you need to define such type separately:

```haskell
data Castle = Castle 
    { castleName :: String
    , castleWall :: Bool 
    }
```

And in the city we will have `Maybe Castle`.

Another type that we need is the enum for Church or Library.
The way you introduce  t inside the City sum type would not work as you expect, as it creates two separate constructors of the `City` data type, and they won't be able to exist at the same time with, for example Castle. So you need to specify this type separately too:

```haskell
data ChurchOrLibrary 
    = Church
    | Library
```

and as you notice to represent house you need a enum.
House should also be a sum type that shows how many people it could accommodate 🙂 

```haskell
data House 
    = ForOne
    | ForTwo
    | ForThree
    | ForFour
```

After all that is done, Then let's build the city:

```haskell
data City = City
    -- Optional castle with a __name__ (as 'String')
    -- Wall, but only if the city has a castle
    { cCastle :: Castle -- from the above
    -- OR you can do it with your type
    -- cCastle :: Maybe Castle
    -- Church or library but not both
    , cChurchOrLibrary :: ChurchOrLibrary
    -- ny number of houses. Each house has one, two, three or four __people__ inside.
    , mcHouse :: [House]
    }
```

And after that, you would be able to implement other functions. Let me know if this helps!

**Review by vrom911 — APPROVED** [review]
Great one! 🤗 

I tried to answer your questions. Hope it is helpful. Let me know if I can clarify more 🙂

**Comment by JosephTLucas** [comment]
Very helpful feedback.  Thank you! No need to recheck yet, still working!
